lorien/Misc/Parser.gd

```diff
@@ -0,0 +1,197 @@
+	var subsymbols: Array
+	var value
+	
+	func _init(_name: String, _last_position: int, _subsymbols: Array, _value).():
+		name = _name
+		last_position = _last_position
+		subsymbols = _subsymbols
+		value = _value
+	
+	func find_first_subsymbol(symbol_name: String):
+		for st in subsymbols:
+			if st.name == symbol_name:
+				return st
+		return null
+	
+	func _to_string():
+		var subsymbol_strings = PoolStringArray([])
+		for st in subsymbols:
+			subsymbol_strings.append(str(st))
+		return "<T name={name}{value} [{subsymbols}]>".format({
+			"name": name,
+			"value": (" '%s'" % value) if value else "",
+			"subsymbols": subsymbol_strings.join(", ")
+		})
+
+# -------------------------------------------------------------------------------------------------
+class GrammarElement:
+	func parse(s: String):
+		pass
+	
+
+# -------------------------------------------------------------------------------------------------
+class Grammar:
+	extends GrammarElement
+	
+	var replace_name = null
+	var elements: Array = []
+	
+	func parse(s: String):
+		var best = null
+		for e in elements:
+			var c = e.parse(s)
+			if c is ParsedSymbol:
+				if (best == null) || (best.last_position < c.last_position):
+					best = c
+		return best
+
+# -------------------------------------------------------------------------------------------------
+class GrammarSequence:
+	extends GrammarElement
+
+	var name: String
+	var elements: Array
+	var flatten_same_name := false
+	
+	func _init(_name: String, _elements: Array = [], _flatten_same_name = false):
+		name = _name
+		elements = _elements
+		flatten_same_name = _flatten_same_name
+
+	func parse(s: String):
+		var position := 0
+		var symbols := []
+		
+		for e in elements:
+			var t = e.parse(s.substr(position))
+			if !t:
+				return null
+			t = t as ParsedSymbol
+			position += t.last_position
+			t.last_position = position
+			symbols.append(t)
+		
+		if flatten_same_name:
+			var new_symbols := []
+			for t in symbols:
+				if t.name == name:
+					new_symbols.append_array(t.subsymbols)
+				else:
+					new_symbols.append(t)
+			symbols = new_symbols
+			
+		return ParsedSymbol.new(name, position, symbols, null)
+
+# -------------------------------------------------------------------------------------------------
+class GrammarLiteral:
+	extends GrammarElement
+	
+	var name: String
+	var value: String
+	var ignore_whitespace := true
+	
+	func _init(_name: String, _value = null):
+		if _value == null:
+			_value = _name
+		name = _name
+		value = _value
+	
+	func parse(s: String):
+		var test_in := s
+		if ignore_whitespace:
+			test_in = test_in.strip_edges(true, false)
+		if !test_in.begins_with(self.value):
+			return null
+		
+		return ParsedSymbol.new(name, s.find(value) + len(value), [], value)
+		
+# -------------------------------------------------------------------------------------------------
+class GrammarRegexMatch:
+	extends GrammarElement
+	
+	var name: String
+	var regex := RegEx.new()
```

Comment:
regex is initialized initialized in the constructor as well. You can just use a normal type hint here.